torch/_decomp/__init__.py

```diff
@@ -250,13 +252,172 @@ def remove_decompositions(
 import torch._refs
 
 
+# Our strategy for deciding if we can preserve a op is following:
+# 1. The op should be known statically that it is functional
+# 2. If it is maybe aliasing, we decompose because we must know if an op
+#    is mutating or aliasing.
+# TODO (tmanlaibaatar) make this utility function and share it with functional_tensor
+# decomp part. (https://github.com/pytorch/pytorch/issues/129431)
+def _check_valid_to_preserve(op_overload):
+    if op_overload in FunctionalTensor.maybe_aliasing_or_mutating_ops:
+        return False
+    if op_overload in FunctionalTensor.metadata_fns:
+        return False
+
+    alias_info = len(
+        [i for i in op_overload._schema.arguments if i.alias_info is not None]
+    )
+
+    is_mutating_or_aliasing = alias_info != 0 or op_overload._schema.is_mutable
+
+    if is_mutating_or_aliasing:
+        return False
+
+    if not torch._C._dispatch_has_kernel(op_overload.name()):
+        return False
+
+    return True
+
+
+def _is_cia_op(op: "OpOverload") -> bool:
+    return (
+        torch._C._dispatch_has_kernel_for_dispatch_key(
+            op.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+        )
+        or torch._C.DispatchKey.CompositeImplicitAutograd in op.py_kernels
+    )
+
+
+def _is_special_op_to_decompose(*args, **kwargs):
+    kernel = kwargs["kernel"]
+    del kwargs["kernel"]
+    # Can't call kernel.decompose due to infinite recursion as
+    # we register this kernel to py_impl directly
+    dk = torch._C.DispatchKey.CompositeImplicitAutograd
+    if torch._C._dispatch_has_kernel_for_dispatch_key(
+        kernel.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+    ):
+        return kernel._op_dk(dk, *args, **kwargs)
+    else:
+        return NotImplemented
+
+
+def _is_special_op_to_preserve(*args, **kwargs):
+    return NotImplemented
+
+
+def _collect_all_valid_cia_ops():
+    cia_ops = torch._C._dispatch_get_registrations_for_dispatch_key(
+        "CompositeImplicitAutograd"
+    )
+    # Ignore quantized namespace ops
+    cia_ops = [name[6:] for name in cia_ops if name.startswith("aten::")]
+    # Materialize all CIA ops first
+    for op in cia_ops:
+        split_list = op.split(".")
+        # Sometime overload could be missing
+        assert len(split_list) == 1 or len(split_list) == 2
+        op_name = split_list[0]
+        op_overload_name = "default"
+        if len(split_list) == 2:
+            op_overload_name = split_list[1]
+
+        _ = getattr(getattr(torch.ops.aten, op_name), op_overload_name)
+
+    cia_ops = set()
+    for op in torch.ops.aten:
```

Comment:
maybe worth a quick comment, since this feels a bit subtle:

(1) you needed to do the `getattr()` stuff above because `torch.ops.aten` is lazily populated, so you aren't guaranteed that just looping over it will give you every CIA op

(2) you **also** need to manually loop over ever aten op here, because you need to query for CIA decompositions that live in both C++, and in the python dispatcher